add .withOutputParameter to expected call

jacob-keller opened the pull request:
When writing mock expectations, a mock implementation may have an output
parameter. In some tests it may be that the output parameter remains
unmodified.

To write an expectation which has an output parameter that is unmodified
you can use the following code:

    .withOutputParameterReturning("name", NULL, 0)

This is somewhat confusing, so add a shorthand

    .withOutputParameter("name")

This sets up the mock function to expect the "name" output parameter but
leave it unmodified.

jacob-keller commented:
Perhaps a better method might be something like ".ignoreParameter" or ".ignoreOutputParameter", which is something else that could be useful for tests which want to ignore a specific parameter but catch the addition of new parameters in the future (vs using .ignoreOtherParameters).

That looks more difficult to implement so I didn't bother.

basvodde commented:
Could it at least be called "withUnmodifiedOutputParameter or something like that?

jacob-keller commented:
I like that. I'm also thinking of the ability to expect a regular parameter but with an unknown value as well, so I will look into adding that and updating the test suite.

basvodde commented:
Will you change the name still?

jacob-keller commented:
Yes, i am planning on updating this pull request. I didn't manage to get to
it today.

On Mon, Feb 10, 2020, 9:07 PM Bas Vodde <notifications@github.com> wrote:

> Will you change the name still?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/cpputest/cpputest/pull/1276?email_source=notifications&email_token=AAGVRXZEX36W6AJO4Q2MUW3RCIW7NA5CNFSM4KRR6NU2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOELLHH5I#issuecomment-584479733>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAGVRXYGDRLJEVBVR33MY43RCIW7NANCNFSM4KRR6NUQ>
> .
>

jacob-keller commented:
I renamed everything to be .withUnmodifiedOutputParameter() and added tests.

The type is still "const void *". I considered adding a custom type like "<none>" or "<empty>" to represent this, but it doesn't seem to gain us much.

I also thought about going the route of adding ".ignoreParameter(name)" but that seems difficult to get right.


For implementing something like ".expectParameterWithAnyValue(name)", I think we would have to add a custom type that represents that fact.

jacob-keller commented:
Added a test case for the _c implementation.

basvodde commented on tests/CppUTestExt/MockParameterTest.cpp:
```diff
@@ -634,6 +634,20 @@ TEST(MockParameterTest, outputParameterSucceeds)
 }
 
+TEST(MockParameterTest, unmodifiedOutputParameterSucceeds)
+{
```
This test is copy-pasted. The retval variable doesn't make a lot of sense in this test. Could you please clean this up and also check the rest of the code that this didn't happen?

Also, remove comments please :)

basvodde commented:
Ok, I was thinking of unwanted side-effects of this, but I cannot find any. So, it'll be good like this. Thanks!

jacob-keller commented:
Hmm seems like something is still not working right.

jacob-keller commented:
Looks like only a DOS build is failing with some sort of allocation issue..

basvodde commented:
Yeah, the DOS build needs to be fixed or disabled. It is on the todo list. Now it just causes confusion.
